### PR TITLE
Modify wheel#unzip to preserve file execute permissions.

### LINKED ISSUE #45: Including PyPl libraries with executable binaries
I am trying to include a the [Modin](https://github.com/modin-project/modin) project as a dependency. This project uses [Ray](https://github.com/ray-project/ray) which in turn uses Redis.

When executing the py_test rule that exercises this code, I get the following permissions error:

```
PermissionError: [Errno 13] Permission denied: '/private/var/tmp/_bazel_joshuaharrison/29b3465cefdffc832f5dd334d22b0ea7/execroot/Crestone/bazel-out/darwin-fastbuild/bin/strategies/example/test_strategy_test.runfiles/pip/pypi__ray/ray/core/src/ray/thirdparty/redis/src/redis-server'
```

This file path is a symlink to /private/var/tmp/_bazel_joshuaharrison/29b3465cefdffc832f5dd334d22b0ea7/external/pip/pypi__ray/ray/core/src/ray/thirdparty/redis/src/redis-server which has the following permissions:

```
-rw-r--r--  1 joshuaharrison  wheel  1343096 Jun  9 16:10 /private/var/tmp/_bazel_joshuaharrison/29b3465cefdffc832f5dd334d22b0ea7/external/pip/pypi__ray/ray/core/src/ray/thirdparty/redis/src/redis-server
```

If I manually chmod this file into being executable, I get a separate error about another non-executable file. If I keep fixing these errors in the same manner, the test will eventually run and pass.

I understand that this is not an issue specific to your project, but was hoping for some help in figuring out a possible solution. For example, I suppose I could make all files executable in [extract_wheel](https://github.com/dillon-giacoppo/rules_python_external/blob/master/extract_wheels/lib/bazel.py#L119)?

Thoughts?

### PR DESCRIPTION
**Pull request by JoshuaCrestone**
Resolves #45

### CODE REVIEW

File: extract_wheels/lib/wheel.py [hunk 1]
```diff
@@ -1,13 +1,31 @@
...
 import zipfile
-from typing import Dict, Optional, Set
+from typing import Dict, Optional, Set, Text, Union
 
 import pkg_resources
 import pkginfo
 
 
+def current_umask():
+    # type: () -> int
```
Comment (dillon-giacoppo) [inline extract_wheels/lib/wheel.py]: This should be inline type definitions to match existing code:
```python
def current_umask() -> int:
```

File: extract_wheels/lib/wheel.py [hunk 2]
```diff
@@ -1,13 +1,31 @@
...
+    """Get the current umask which involves having to set it temporarily."""
+    mask = os.umask(0)
+    os.umask(mask)
+    return mask
+
+
+def set_extracted_file_to_default_mode_plus_executable(path):
+    # type: (Union[str, Text]) -> None
```
Comment (dillon-giacoppo) [inline extract_wheels/lib/wheel.py]: No need for this to be `Union[str, Text]`, just `str` should be fine.

File: extract_wheels/lib/wheel.py [hunk 3]
```diff
@@ -43,6 +61,15 @@ def dependencies(self, extras_requested: Optional[Set[str]] = None) -> Set[str]:
...
         with zipfile.ZipFile(self.path, "r") as whl:
             whl.extractall(directory)
+            # The following logic is borrowed from Pip:
+            # https://github.com/pypa/pip/blob/cc48c07b64f338ac5e347d90f6cb4efc22ed0d0b/src/pip/_internal/utils/unpacking.py#L240
+            for info in whl.infolist():
+                mode = info.external_attr >> 16
+                # if mode and regular file and any execute permissions for
+                # user/group/world?
+                if mode and stat.S_ISREG(mode) and mode & 0o111:
+                    name = os.path.join(directory, info.filename)
```
Comment (dillon-giacoppo) [inline extract_wheels/lib/wheel.py]: The upstream code has an if/else guard to guarantee that the file is not a directory, i.e 
```python
if fn.endswith('/') or fn.endswith('\\')
```

I think this behaviour should be preserved.

**Review by dillon-giacoppo — CHANGES_REQUESTED** [review]
Thanks for the patch. A couple style issues and it would be good to check if the file is a directory like Pip does. Once everything looks good, please run the file through the Black auto formatter.

Also can you confirm that `ray` installs correctly with the patch.

File: extract_wheels/lib/wheel.py [hunk 4]
```diff
@@ -1,13 +1,29 @@
 """Utility class to inspect an extracted wheel directory"""
 import glob
 import os
+import stat
 import zipfile
-from typing import Dict, Optional, Set
+from typing import Dict, Optional, Set, Text, Union
```
Comment (dillon-giacoppo) [inline extract_wheels/lib/wheel.py]: nit: these are now unused

**Review by dillon-giacoppo — APPROVED** [review]
LGTM once the unused imports are cleaned up. Thanks for the contribution 🎉

File: extract_wheels/lib/wheel.py (hunk 4 above)
Comment (JoshuaCrestone) [reply extract_wheels/lib/wheel.py]: Ahh - sorry. Done.

**Pull request merged**